yo1k/sandbox/exercises/CSC_algorithms/sorting.py

```diff
@@ -0,0 +1,131 @@
+from collections.abc import MutableSequence
+from typing import Any, Optional
+
+
+def swap(array: MutableSequence[Any], frst_idx: int, snd_idx: int) -> MutableSequence[Any]:
+    """Swaps to elements in a list."""
+    array[frst_idx], array[snd_idx] = array[snd_idx], array[frst_idx]
+    return array
+
+
+def selection_sort(array: MutableSequence[Any]) -> MutableSequence[Any]:
+    """Sorting by selection.
+
+    Algorithm in-place, complexity: T(n) = O(n**2).
+    """
+    size: int = len(array)
+    for i in range(size):
+        min_idx: int = -1
+        for j in range(i, size):
+            if min_idx == -1 or array[j] <= array[min_idx]:
+                min_idx = j
+        swap(array, min_idx, i)
+    return array
+
+
+def insertion_sort(array: MutableSequence[Any]) -> MutableSequence[Any]:
+    """Sorting by inserts.
+
+    Algorithm in-place, complexity: T(n) = O(n**2).
+    """
+    for i in range(len(array)):
+        j: int = i
+        while j > 0 and array[j - 1] > array[j]:
+            swap(array, j-1, j)
+            j -= 1
+    return array
+
+
+def merge_sorted_arrays(
+        array_left: MutableSequence[Any],
+        array_right: MutableSequence[Any],
+        merged_array: Optional[MutableSequence[Any]] = None) -> MutableSequence[Any]:
+    """Merges two sorted arrays in ascending order."""
+    merged_array = [] if merged_array is None else merged_array
+    l_idx: int = 0
+    r_idx: int = 0
+    l_end: int = len(array_left)
+    r_end: int = len(array_right)
+
+    while l_idx < l_end and r_idx < r_end:
+        if array_left[l_idx] <= array_right[r_idx]:
+            merged_array.append(array_left[l_idx])
+            l_idx += 1
+        else:
+            merged_array.append(array_right[r_idx])
+            r_idx += 1
+    while l_idx < l_end:
+        merged_array.append(array_left[l_idx])
+        l_idx += 1
+    while r_idx < r_end:
+        merged_array.append(array_right[r_idx])
+        r_idx += 1
+    return merged_array
+
+
+def merge_sort(array: MutableSequence[Any]) -> MutableSequence[Any]:
+    """Mergesort.
+
+    This algorithm has complexity O(n*log(n)) on condition
+    when we do not take into account overheads of creating new arrays in the process of sorting,
+    only comparison operation."""
```

Comment:
This part of the sentence is weird, I propose removing it.

Reply:
Done.

Reply:
Done.